File: postgresql/system/data_type.pl
#!/usr/bin/perl
# Copyright (C) 2022 The go-postgresql Authors. All rights reserved.
# Use of this source code is governed by a BSD-style
# license that can be found in the LICENSE file.

print<<HEADER;
// Copyright (C) 2022 The go-postgresql Authors. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package system

type DataType struct {
	name string
	oid  OID
	size int
}

// DataType represents a PostgreSQL data type.
var dataTypes = map[OID]*DataType{}

// GetDataType returns a data type for the specified OID.
func GetDataType(oid OID) (*DataType, error) {
	dt, ok := dataTypes[oid]
	if !ok {
		return nil, newDataTypeNotFound(oid)
	}
	return dt, nil
}

func newDataType(name string, oid OID, size int) *DataType {
	dt := &DataType{
		name: name,
		oid:  oid,
		size: size,
	}
	return dt
}

// Name returns the data type name.
func (dt *DataType) Name() string {
	return dt.name
}

// OID returns the data type OID.
func (dt *DataType) OID() OID {
	return dt.oid
}

// Size returns the data type size.
func (dt *DataType) Size() int {
	return dt.size
}

// FormatCodeFrom returns a format code from the specified data type.
func (dt *DataType) FormatCode() FormatCode {
	switch dt.oid {
	case Bytea:
		return BinaryFormat
	}
	return TextFormat
}

func init() { //nolint: maintidx
HEADER

my $pg_type_file = "res/pg_type.csv";
open(IN, $pg_type_file) or die "Failed to open $pg_type_file: $!";
my $line_no = 0;
my $name_idx = -1;
my $oid_idx = -1;
my $typ_len_idx = -1;
while(<IN>){
  $line_no++;
  chomp($_);
  my @row = split(/,/, $_, -1);
  if ($line_no <= 1) {
    for (my $i = 0; $i < scalar(@row); $i++) {
      if ($row[$i] eq "typname") {
        $name_idx = $i;
      }
      if ($row[$i] eq "oid") {
        $oid_idx = $i;
      }
      if ($row[$i] eq "typlen") {
        $typ_len_idx = $i;
      }      
    }
    if ($name_idx < 0) {
      die "Failed to find typname column";
    }
    if ($oid_idx < 0) {
      die "Failed to find oid column";
    }
    if ($typ_len_idx < 0) {
      die "Failed to find typlen column";
    }
    next;
  }
  my $name = $row[$name_idx];
  my $oid = $row[$oid_idx];
  my $size = $row[$typ_len_idx];

  if ($name =~ /^_/){
    next;
  }
  # check nil rows
  if (length($name)<=0 || length($oid)<=0 || length($size)<=0) {
    next;
  }
  # UpperCase
  # $name =~ s/([a-z])/\u$1/g;
  # CamelCase
  $name =~ s/_([a-z])/\u$1/g;
  $name =~ s/^([a-z])/\u$1/;
  # stylecheck
  my @patterns = ('Sql', 'Xml', 'Json', 'Uuid', 'Db', 'Ts', 'Acl');
  my @replacements = ('SQL', 'XML', 'JSON', 'UUID', 'DB', 'TS', 'ACL');
  for (my $i = 0; $i < scalar(@patterns); $i++) {
    if ($name =~ /$patterns[$i]/) {
      $name =~ s/$patterns[$i]/$replacements[$i]/g;
    }
  }
  print "\tdataTypes[" . $oid . "] = newDataType(\"" . $name . "\", " . $oid . ", " . $size . ")\n";  
}
close(IN);

print<<FOTTER;
}
FOTTER

File: postgresql/query/data_type.go
// Copyright (C) 2019 The go-postgresql Authors. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package query

import (
	"fmt"

	"github.com/cybergarage/go-postgresql/postgresql/errors"
	"github.com/cybergarage/go-postgresql/postgresql/system"
	"github.com/cybergarage/go-sqlparser/sql/query"
)

// PostgreSQL: Documentation: 16: Chapter 8. Data s
// https://www.postgresql.org/docs/16/datatype.html

// DataType represents a data type.
type DataType = system.DataType

// ObjectID represents a object identifier.
type ObjectID = system.ObjectID

// NewDataTypeFrom returns a data type from the specified query data type.
func NewDataTypeFrom(t query.DataType) (*DataType, error) {
	objectID, err := NewObjectIDFrom(t)
	if err != nil {
		return nil, err
	}
	return system.NewDataTypeFromObjectID(objectID)
}

// NewObjectIDFrom returns a data type from the specified query data type.
func NewObjectIDFrom(t query.DataType) (ObjectID, error) {
	switch t {
	case query.BigIntType:
		return system.Int8, nil
	case query.BinaryType:
		return system.Bytea, nil
	case query.BitType:
		return system.Bit, nil
	case query.BlobType:
		return system.Bytea, nil
	case query.BooleanType:
		return system.Bool, nil
	case query.CharType:
		return system.Char, nil
	case query.CharacterType:
		return system.Varchar, nil
	// case query.ClobData:
	// 	return
	case query.DateType:
		return system.Date, nil
	// case query.DecimalData:
	// 	return
	case query.DoubleType:
		return system.Float8, nil
	case query.FloatType:
		return system.Float4, nil
	case query.IntType:
		return system.Int4, nil
	case query.IntegerType:
		return system.Int4, nil
	case query.LongBlobType:
		return system.Bytea, nil
	case query.LongTextType:
		return system.Text, nil
	case query.MediumBlobType:
		return system.Bytea, nil
	// case query.MediumIntData:
	// 	return
	case query.MediumTextType:
		return system.Text, nil
	// case query.NumericData:
	// 	return
	// case query.RealData:
	// 	return
	// case query.SetData:
	// 	return
	case query.SmallIntType:
		return system.Int2, nil
	case query.TextType:
		return system.Text, nil
	case query.TimeType:
		return system.Time, nil
	case query.TimeStampType:
		return system.Timestamp, nil
	case query.DateTimeType:
		return system.Timestamp, nil
	case query.TinyBlobType:
		return system.Bytea, nil
	// case query.TinyIntData:
	// 	return
	case query.TinyTextType:
		return system.Text, nil
	case query.VarBinaryType:
		return system.Bytea, nil
	case query.VarCharType:
		return system.Varchar, nil
	case query.VarCharacterType:
		return system.Varchar, nil
		// case query.YearData:
		// 	return
	}
	return 0, fmt.Errorf("data type (%s) %w", t, errors.ErrNotSupported)
}
